src/coverlet.core/Helpers/InstrumentationHelper.cs

```diff
@@ -189,25 +182,57 @@ public bool PortablePdbHasLocalSource(string module, out string firstNotFoundDoc
+        private bool MatchDocumentsWithSources(string module, AssemblySearchType excludeAssembliesWithoutSources,
+            MetadataReader metadataReader)
         {
-            foreach (DocumentHandle docHandle in metadataReader.Documents)
+            if (excludeAssembliesWithoutSources.Equals(AssemblySearchType.MissingAll))
             {
-                Document document = metadataReader.GetDocument(docHandle);
-                string docName = _sourceRootTranslator.ResolveFilePath(metadataReader.GetString(document.Name));
-                Guid languageGuid = metadataReader.GetGuid(document.Language);
-
-                // We verify all docs and return false if not all are present in local
-                // We could have false negative if doc is not a source
-                // Btw check for all possible extension could be weak approach
-                // We exlude from the check the autogenerated source file(i.e. source generators)
-                // We exclude special F# construct https://github.com/coverlet-coverage/coverlet/issues/1145
-                if (!_fileSystem.Exists(docName) && !IsGeneratedDocumentName(docName) &&
-                    !IsUnknownModuleInFSharpAssembly(languageGuid, docName))
+                bool anyDocumentMatches = MatchDocumentsWithSourcesMissingAll(metadataReader);
+                if (!anyDocumentMatches)
                 {
-                    return (false, docName);
+                    _logger.LogVerbose($"Unable to instrument module: {module}, pdb without any local source files");
+                    return false;
                 }
             }
+
+            if (excludeAssembliesWithoutSources.Equals(AssemblySearchType.MissingAny))
+            {
+                (bool allDocumentsMatch, string notFoundDocument) = MatchDocumentsWithSourcesMissingAny(metadataReader);
+
+                if (!allDocumentsMatch)
+                {
+                    _logger.LogVerbose(
+                        $"Unable to instrument module: {module}, pdb without local source files, [{FileSystem.EscapeFileName(notFoundDocument)}]");
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private IEnumerable<(string documentName, bool documentExists)> DocumentSourceMap(MetadataReader metadataReader)
+        {
+            return metadataReader.Documents.ToList().Select(docHandle =>
+            {
+                Document document = metadataReader.GetDocument(docHandle);
+                string docName = _sourceRootTranslator.ResolveFilePath(metadataReader.GetString(document.Name));
+                return (docName, _fileSystem.Exists(docName));
+            });
+        }
+
+        private bool MatchDocumentsWithSourcesMissingAll(MetadataReader metadataReader)
+        {
+            var documentSourceMap = DocumentSourceMap(metadataReader).ToList();
+            return documentSourceMap.Any(x => x.documentExists.Equals(true));
```

Comment:
Minor detail, this could work on the enumeration directly, but at the price of making debugging harder:

```suggestion
            return DocumentSourceMap(metadataReader).Any(x => x.documentExists);
```

Reply:
done